Add more information to BOOTSTRAP.adoc

sabine opened the pull request:
This PR adds more information about the bootstrap process regarding
1. when exactly to commit for upstreaming, and
2. changes to primitives

In https://github.com/ocaml/ocaml/pull/11264, I struggled with the bootstrap. After questioning @dra27 and investigating the errors I saw during my failing attempts, I compiled some answers that would have helped me to avoid confusion.

sabine committed:
Make BOOTSTRAP.adoc more precise

1. gives context and clarifies the steps of the bootstrap process.
   In particular: when exactly to commit when upstreaming.
2. adds some background information on primitives that may be helpful
   in understanding why changes to primitives need a bootstrap
3. add step by step instructions for removing a primitive and
   clarifies existing instructions for renaming

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

shindere commented:
Many thanks for taking care about this file, @sabine.

You could improve the commit message: at the moment it does not mention
what it modifies. And, I personnally like commit messages to start with
a real title, which means its first letter should be in upper-case.

One possible first line for your commit could be:

```
Make BOOTSTRAP.adoc more precise
```

Regarding the text of the change itself, I find it clear and nice. I am
just suggesting a few improvements of the wording:

`Indicate in the message of the commit`: `Indicate in the commit
message`

`To create the primitives table (used by both the interpreter and by the
+compiler)`: I think you could remove the second occurrence of `by`. I
would also remove the comma after the parentheses.

`a script scans for the tag `CAMLprim` in +the C files of the runtime
and collects the names of these functions`: two independent suggestions
here: 1. Move the for the tag `CAMLprim` bit at the end and remove the
`in`word. 2. Replace `tag` with `C preprocessor macro` which conveys
more information.

About the phrase `Commit the bootstrap` that you use twice I think. How
about "commit the updated (bootstrapped) binaries?

Regarding `5. Remove the old primitive stub`. Earlier in the document,
you distinguished the uses of a primitive in the OCaml world and the
implementatin of primitives, in the C world. I guess that `stub` refers
to the implementation in C but this is something you may make more
explicit e.g. by introducing the word `stub` when you first mention
these two parts of promitives. You could also of course not mention stub
at all and stick to your first terminology.

sabine commented:
@shindere Great feedback! Thank you very much, I improved the doc accordingly.

shindere commented:
Many thanks @sabine for having taken the comments into account and
apologies for the time it took me to review!

I would like to suggest a potential improvement of the explanation
regarding the primitives.

The sentence `The OCaml compiler internally defines so-called
"primitives".` talks aobut the primitives from the point of view of the
compiler. IIt would seem clearer to me to introduce them from the point
of view of an OCmal program. Perhaps something like

```
The execution of an OCaml program relies on so-called "primitives".
```

```
A primitive is +more or less a C function of the runtime that is lifted
into OCaml by means of +the usual OCaml-C FFI conventions.
```

Why do you write "more or less"?Also I'd suggest to not mention that the
functions are written in C because, from my point of view, this is just
an implementation detail and the fact that these primitive functions are
written in C is not really relevant to help understanding exactly what
they are.

`that is lifted into OCaml by means of the usual OCaml-C FFI
conventions`: this bit looked a bit overcomplicated to me. How about
saying that the fucntion will be called from OCaml code and that it must
hence obey the usual FFI rules?

Finally on this, the example you give that primitives are used to
implement functions from the stdlib looks a bit accidental to me. Yes,
it's true, but I am not sure whether that's the most important use of
primitives. I'd rather give as an example primitives that implement
features of the language itself, for instance testing equality or
something like that.

Also, the quoting around CAMLprim looks slightly wrong to me, I think
the double quotes could be removed, right?

`The resulting function names`: how about the *collected* funciton
names?

Nitpick: you use `subsequently` several times in the text, perhaps you
could replace at least some of its occurrences just by `then`?

```
Every OCaml bytecode executable file contains a section `PRIMS` which
contains a +sequence of all the names of the primitives it was compiled
with (see +`bytecomp/bytelink.ml`).
```

suggestion: contains a `PRIMS` section listing all the primitives this
program relies on.

One last general suggestion: it feels to me that two styles are used in
the docuemnt: sometimes it's the infinitive form, sometimes it's a more
personal one (when you remove). I'd suggest to be consistent and always
use the same form. My preference would be for the infinitive one.

shindere commented:
@sabine have you seen

https://github.com/ocaml/ocaml/pull/11264#issuecomment-1137249260?

Just in case it has not already been taken into account, this PR seems a
good time to do so.

sabine commented:
@shindere Very good points and thanks for the actionable advice! I force pushed the improved document.

https://github.com/ocaml/ocaml/pull/11264#issuecomment-1137249260 is the reason I made this attempt at clarifying, so indeed it's covered.

shindere commented:
Many thanks, @sabine, for your prompt and positive response.

I am going to send a last series of suggestions and then, I think, it
will be fine and if it's not we will still merge your PR and I can
always submit another one on top of it.

When the different binaries that are successfully built are mentionned,
at the beginning of the document, it may help people to clearly state
their names (ocamlc in the top-level directory, etc.).

```
1. Make sure to start with a clean source tree (e.g. check with `git status`)
```

Suggestion: (e.g. check with `git status` that your work tree is clean;
you may want to start in a new branch too, check it out with `git
checkout -b new-branch`).

```
./boot/ocamlrun ./ocamlc -I ./stdlib foo.ml
```

I didn't try but my guess is that this command wouldn't work because
it's missing the `-nostdlib` option. I know this was not added by you
but if you feel like doing so, would you mind giving it a try and either
confirm that the present line works, or add the `-nostdlib` flag,
please?

Also, regarding the same line, you could add a `-o hello.exe` to it and
modify the subsequent command-line, because in its current state, this
would work only on Unix (because of the hypothesis about the name of the
executable being a.out).

Regarding the section `Adding, removing and renaming primitives`: it
seems to me it would be nice to make explicit where this belongs, in
terms of the steps that have been previously listed. Si I think the
section relates to step 4.

Just before the `== Adding a primitive`section, how about adding one
paragraph explaining the primitives from the point of view of the
compiler?

What I mean here is that the bytecod OCaml compiler itself actually
depends on two sets of primitives. On the one hand, there is the set of
primitives it requres to run, so this corresponds to what is written
just above for any OCaml bytecode program. On the other hand, there is
the set of primitives it may call in the programs it generates. In
theory, these two seets of primitives do not have to be connected with
each other, although they are in practice very close, if not the same.

With these explanations, it feels that what follows wil become even
clearer.

```
1. Start from a stable state
```

Suggestion: at the end of this line, add `(see step 3 above)`, and then
you can perhaps remove the commands?

```
2. Remove primitive from `.ml/.mli` files
```

Suggestion:

2. Remove the primitive from OCaml source files (`.ml`,  `.mli`,
`.mll`, `.mly` etc.)

```
3. Rebuild the core system.
```

Can you say here to which of the steps above this is referring?

The same remark goes for the other steps and perhaps if this refers to
consecutive steps of the series above you can just summarize and say
`repeat steps x to y above`?

One general remark is that the document feels very clear now and
helpful! So, thanks a lot, @sabine!

sabine commented:
@shindere Great! I addressed all these comments.

I moved the recommendation of `make distclean` to the general bootstrap procedure and accordingly simplified the specialized bootstrap procedures for primitives to refer to the general bootstrap procedure. Turns out, the specialized bootstrap procedures replace steps 4 and 5 of the general bootstrap procedure. I named the steps of the specialized bootstrap procedures with letters, to make it easier to communicate when someone is stuck trying to bootstrap.

It looks to me like steps 4 and 5, in the case of renaming or removing primitives, are actually intertwined with each other in the sense that we first change some code, then build, then change more code (by removing a C primitive).

Generally: Would it make sense to ask people to commit after `make coreall` and have the commit message mention that this is the place to do `make coreall`? And similarly, ask people to mark the commit on which `make bootstrap` should be done?

shindere commented:
Hey @sabine, great work!

sabine (2022/06/23 03:25 -0700):
> @shindere Great! I addressed all these comments.
>
> I moved the recommendation of `make distclean` to the general
> bootstrap procedure and accordingly simplified the specialized
> bootstrap procedures for primitives to refer to the general bootstrap
> procedure. Turns out, the specialized bootstrap procedures replace
> steps 4 and 5 of the general bootstrap procedure. I named the steps of
> the specialized bootstrap procedures with letters, to make it easier
> to communicate when someone is stuck trying to bootstrap.

All this makes a lot of sense to me. Many thanks for having invested the
time and energy in it!

> It looks to me like steps 4 and 5, in the case of renaming or removing primitives, are actually intertwined with each other in the sense that we first change some code, then build, then change more code (by removing a C primitive).
>
> Generally: Would it make sense to ask people to commit after `make
> coreall` and have the commit message mention that this is the place to
> do `make coreall`? And similarly, ask people to mark the commit on
> which `make bootstrap` should be done?

My impression is that people who willhave to take care with these things
already know how to handle them so I'd tend to say `no`, but other core
developers may have different opinions on that. @gasche or @dra27, for
instance?
